runtime/major_gc.c

```diff
@@ -148,10 +148,61 @@ static void realloc_gray_vals (void)
   }
 }
 
+#ifdef NATIVE_CODE_AND_NO_NAKED_POINTERS
+
+static int is_pointer_safe (value v, value *p)
+{
+  volatile header_t h;
+  tag_t t;
+
+  Caml_state->checking_pointer_pc = &&on_segfault;
+  h = Hd_val(v);
```

Comment:
This, Sir, is a bold use of GCC's first-class labels, but I kind of like it.

The more conservative approach would `sigsetjmp` here and `siglongjmp` out of the signal handler for SIGSEGV, but maybe we found out previously that it doesn't work.

I'm just hoping the optimizer understands that unpredictable control flow can occur.

Reply:
The optimizer doesn't if the declaration `header_t h` isn't marked volatile. I haven't tried `sigsetjmp/siglongjmp`. I shall give that a try shortly.

Reply:
I'm afraid you'll find `sigsetjmp` and `siglongjmp` too costly (they work with signal masks via system calls?).  The current code is very efficient; all we need it to get it to work reliably despite GCC and Clang optimizations...

Reply:
You are right. `sigsetjmp` and `siglongjmp` is quite slow. I tried `opam install frama-c -y` on a fresh switch. With GCC first-class labels, it takes `3m30.726s`. With `sigsetjmp/siglongjmp` it takes `16m2.556s`.